### DISCUSSION TITLE
Where do the function invocations from ` ___netlify-odb-handler` come from?

### DISCUSSION THREAD

**Discussion by remarcable** [opennextjs/opennextjs-netlify#1311] [Q&A]
Hello there (@ascorbic @ericapisani),
we recently switched building a Next.js app on Netlify from using `next build && next export` to `next build` only. This generated three functions, among them ` ___netlify-odb-handler`.

I saw in the usage tab in the function settings, that in the last ten days, we used about 70% of the function invocations already. The number of calls to the function surprised me because I thought that all pages were statically generated and there shouldn't be any need to call the function to build any page. Investigating this further, I noticed in the logs that almost 100% of the calls were made to ` ___netlify-odb-handler`.

The logs look like this:
```
Apr 19, 11:33:56 AM: 74e548ea INFO   [GET] /.well-known/assetlinks.json (ODB)
Apr 19, 11:33:56 AM: 74e548ea Duration: 22.44 ms	Memory Usage: 126 MB	
Apr 19, 11:34:45 AM: a5dffe37 INFO   [GET] /.well-known/assetlinks.json (ODB)
Apr 19, 11:34:45 AM: a5dffe37 Duration: 6.71 ms	Memory Usage: 127 MB	
Apr 19, 11:35:10 AM: 8d46d7c2 INFO   [GET] /_next/data/J3BGMwUS5B1t2M6O8Jauy/en/piano-guide/piano-pedals.json (ODB)
Apr 19, 11:35:10 AM: 8d46d7c2 Duration: 6.09 ms	Memory Usage: 127 MB	
Apr 19, 11:35:12 AM: 0d62ae6a INFO   [GET] /apple-touch-icon-120x120-precomposed.png (ODB)
Apr 19, 11:35:12 AM: 0d62ae6a Duration: 7.61 ms	Memory Usage: 127 MB	
Apr 19, 11:35:13 AM: 9521f25f INFO   [GET] /apple-touch-icon-120x120.png (ODB)
Apr 19, 11:35:13 AM: 9521f25f Duration: 6.42 ms	Memory Usage: 128 MB	
Apr 19, 11:35:23 AM: faf90abc INFO   [GET] /_next/data/DN3Cmp8qFvKcaUbLEQSEG/en/piano-hands.json (ODB)
Apr 19, 11:35:23 AM: faf90abc Duration: 5.73 ms	Memory Usage: 128 MB	
Apr 19, 11:37:13 AM: d4e2b809 INFO   [GET] /.well-known/assetlinks.json (ODB)
```

When is `___netlify-odb-handler` invoked? From the logs, it appears it is called whenever a page needs to be built, and therefore in our case whenever a 404 happens (all the paths above don't exist).

What is the reason the 404 page needs to be built every time instead of once? Is there a way to disable building the 404 and just display it statically – or if not, to cache the result?

Thank you very much!

**Comment by remarcable** [comment]
@ascorbic @ericapisani What do you think?

**Comment by ericapisani** [comment]
Hi @remarcable ,

> When is ___netlify-odb-handler invoked?

`___netlify-odb-handler` is invoked for ISR and fallback routes (there's a short summary about which functions handle which routes on the plugin README [here](https://github.com/netlify/netlify-plugin-nextjs#generated-functions)), so your findings based on the logs that the function is called whenever a page needs to be built is correct.

Without seeing your project, I can't speculate as to why the 404 page is being built every time, but what I do know is that NextJS provides a static 404 page by default, and if a custom 404 page is being used within the project, it looks like it needs to be named `pages/404.js` (and is also statically generated at build time). 

There's a brief guide on custom 404 pages [on the Netlify blog](https://www.netlify.com/blog/2020/12/08/making-a-custom-404-page-in-next.js/) and for a more comprehensive guide, here's the relevant docs on [NextJS](https://nextjs.org/docs/advanced-features/custom-error-page).

We also have a [demo site](https://github.com/netlify/netlify-plugin-nextjs/tree/main/demos/default) that contains a custom 404 page that is statically generated in case you want to use it to try and compare the differences between your project and what we have to try and determine why it's not being statically generated within your project.

I hope that helps in your debugging efforts, but if not, I'll need more information in order to investigate further (in which case an issue should be opened in the event that you're encountering a potential bug within the plugin).
